[Mobile] Appium 2 Migration

geriux opened the pull request:
## Related
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6276

## What?
This PR updates Appium to the newest `2.0.0` version, it also replaces the usage of `wd` in favor of `WebdriverIO`.

## Why?
Newer versions of Xcode weren't compatible with Appium 1, so the code needs to be migrated to the latest version to stay up to date with the latest versions available and prevent disruptions in the future with a broken E2E testing environment.

## How?

By completing the following tasks:

### Setup
- [x] Update dependencies to Appium 2 and WebdriverIO
- [x] Update CI jobs to run with Xcode 14 and Appium 2
- [x] Update remote connection
- [x] https://github.com/WordPress/gutenberg/pull/55257
- [x] https://github.com/WordPress/gutenberg/pull/55443

### Tests

- [x] Block insertion https://github.com/WordPress/gutenberg/pull/55167
- [x] Device Actions https://github.com/WordPress/gutenberg/pull/55211
- [x] Drag & Drop blocks https://github.com/WordPress/gutenberg/pull/55254
- [x] Heading - https://github.com/WordPress/gutenberg/pull/55215
- [x] Initial HTML
- [x] Media blocks - https://github.com/WordPress/gutenberg/pull/55222
- [x] Paragraph - https://github.com/WordPress/gutenberg/pull/55215
- [x] Rendering Media Blocks
- [x] Rendering Other Blocks
- [x] Rendering Text Blocks
- [x] Search - https://github.com/WordPress/gutenberg/pull/55170

### WD usage

- [x] https://github.com/WordPress/gutenberg/pull/55541

## Testing Instructions
CI checks should pass in both Gutenberg and Gutenberg Mobile.

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

Commit:
Dependencies update:
- Bumps Appium to version 2.0.0
- Removes wd in favor of WebdriverIO

Commit:
- Adds appium:setup script to install needed drivers for Appium.
- Updates test:e2e:build-wda with the new Appium home path for the xcuitest driver.
- Updates test:e2e:android:local and test:e2e:ios:local to add the newly added appium:setup script to install drivers if needed.

Commit:
Update Android Runner to cache the Appium drivers

Commit:
Update iOS Runner - Bump Xcode to 14.2

Commit:
Update iOS Runner with these changes:

- Move the Launch of the simulator step to the top so it starts booting up the simulator in the background, so it doesn't block the next steps while it's stating.
- Install and cache the Appium drivers.

Commit:
Updates the E2E capabilities settings:

- It removes unneeded values after the migration.
- Bumps the iOS platform version to 16.2.
- Adds specific SauceLabs configuration that sets the Appium version to 2.0.0.
- Adds a prefixKeysWithAppium util to append the "appium:" prefix for the Appium capabilities settings. This is added to simplify the modifications we do for some of the settings in runtime.

Commit:
Updates the E2E serverConfigs file with the updated SauceLabs settings for Appium 2 and WebdriverIO

Commit:
E2E Utils:

- Remove import of wd and set it as null temporarily while we migrate its usage to WebdriverIO

Commit:
E2E Utils:

Update getIOSPlatformVersions to get the desired platform version from the capabilities settings instead of an inline value

Commit:
E2E Utils:

- Update Remote configuration with the new structure for WebdriverIO and Appium 2
- Uses the prefixKeysWithAppium util for the Appium capabilities, we cannot use "appium:options" since that's only available for real devices and we use simulators instead.
- The app capability value was updated to the new storage:filename= prefix this was important to make SauceLabs to work correctly

Commit:
E2E Utils - Update the stopDriver function with:

- Changes sessionID to sessionId
- Removes the hash value to append in the Job URL, these URLs are public so it's not needed to append the auth param.
- Updates driver.quit() to driver.deleteSession()

Commit:
E2E Utils - Updates swipeFromTo to WebdriverIO

Commit:
E2E Utils - Updates isEditorVisible to WebdriverIO

It adds a custom timeout to waitForDisplayed, the simulators in CI are slow so we need to give enough time to find the post title locator.

Commit:
Remove usage of sauceJobStatus, WebdriverIO internally handles this

Commit:
EditorPage - Remove import of wd and keep a null value while we migrate to WebdriverIO

Commit:
EditorPage - Migrate initializeEditor to WebdriverIO

Commit:
Gutenberg Editor Rendering Media Blocks test - Migrate to WebdriverIO

geriux commented on .github/workflows/rnmobile-ios-runner.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
         strategy:
             matrix:
-                xcode: ['13.3.1']
+                xcode: ['14.2']
```
For the `macos-12` runner the latest Xcode 14 version is `14.2` ([docs](https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md#xcode))

geriux commented on .github/workflows/rnmobile-ios-runner.yml:
```diff
@@ -27,9 +27,25 @@ jobs:
+            - name: Switch Xcode version to ${{ matrix.xcode }}
+              run: sudo xcode-select --switch /Applications/Xcode_${{ matrix.xcode }}.app
+
+            - name: Launch simulator
```
I've moved this to the beginning using the `&` so it starts booting up in the background and doesn't block the rest of the steps.

geriux commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -146,38 +158,45 @@ const setupDriver = async () => {
+		logLevel: 'error',
+		capabilities: {
+			platformName: PLATFORM_NAME,
+			...prefixKeysWithAppium( desiredCaps ),
```
At first I was using `appium:options` to pass all of these capabilities, but then I found an issue on SauceLabs when it was trying to trigger real devices instead of simulators. It turns out this option of using `appium:options` is only available for Real devices ([docs](https://docs.saucelabs.com/dev/test-configuration-options/#appiumoptions)).

I'm open to other ideas on how to dynamically rename all capabilities to have the `appium:` prefix.

dcalhoun replied:
This approach seems reasonable to me. 👍🏻

geriux commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -146,38 +158,45 @@ const setupDriver = async () => {
 	if ( ! isLocalEnvironment() ) {
-		const jobID = driver.sessionID;
-
-		const hash = crypto
```
This was causing issues due to the server configuration changes for SauceLabs. The jobs links are public so I don't think we need that auth param.

dcalhoun replied:
🤔 I have less familiarity with this infrastructure. Would it make sense to validate this with Apps Infrastructure?

geriux replied:
After looking at this again I investigated further when it was implemented. This was introduced in this PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/849. It gave me insights into the [SauceLabs Sharable links](https://docs.saucelabs.com/test-results/sharing-test-results/index.html#building-sharable-links).

I checked their latest documentation [about this approach](https://docs.saucelabs.com/test-results/sharing-test-results/#example---nodejs) and I ended up updating the code to keep the existing functionality but with some recent updates from the docs. What do you think about keeping the code as it was (with the recent changes)?

geriux replied:
Here's the commit with the code update https://github.com/WordPress/gutenberg/pull/55166/commits/9db690e1ba017c0d49372384ef65bab90fb3024a

dcalhoun replied:
The new approach looks great and feels more appropriate. Well done!

dcalhoun commented on package.json:
```diff
@@ -166,7 +166,7 @@
 		"ajv": "8.7.1",
 		"ajv-draft-04": "1.0.0",
-		"appium": "1.22.3",
+		"appium": "2.0.0",
```
Is there a reason to avoid the latest [`2.1.3` release](https://github.com/appium/appium/releases/tag/appium%402.1.3)?

geriux replied:
To keep in sync with the version we use in SauceLabs I thought it'd be a good idea to match behaviors between running the tests locally and on SauceLabs.

The latest version they support is `2.0.0` from their [platform configurator tool](https://saucelabs.com/products/platform-configurator#/). They say that if [we have specific needs for an Appium version](https://docs.saucelabs.com/mobile-apps/automated-testing/appium/appium-2-migration/#can-i-provide-my-own-appium-server-version) we can reach out to their support team. I don't think this is currently the case but it's worth keeping it in mind for the future.

dcalhoun commented on packages/react-native-editor/__device-tests__/helpers/serverConfigs.js:
```diff
@@ -4,7 +4,9 @@ exports.local = {
+	key: process.env.SAUCE_ACCESS_KEY,
+	hostname: 'ondemand.us-west-1.saucelabs.com',
+	port: 443,
+	baseUrl: 'wd/hub',
```
Is this retaining the [legacy convention](https://appium.io/docs/en/2.1/guides/migrating-1-to-2/#default-server-base-path)? Is that required?

geriux replied:
I've checked the docs for SauceLabs and they haven't changed the `baseUrl`, my guess is that they haven't fully migrated and are keeping the same URL 🤔. To double check I used their [Platform Configurator](https://saucelabs.com/products/platform-configurator#/) and it's still returning that base URL. We should keep an eye on when they make the change.

dcalhoun commented on packages/react-native-editor/package.json:
```diff
@@ -108,14 +108,15 @@
 		"device-tests-canary": "cross-env NODE_ENV=test jest --forceExit --detectOpenHandles --no-cache --maxWorkers=2 --testPathPattern=@canary --config ./jest_ui.config.js",
 		"device-tests:local": "cross-env NODE_ENV=test jest --runInBand --detectOpenHandles --forceExit --config ./jest_ui.config.js",
 		"device-tests:debug": "cross-env NODE_ENV=test node $NODE_DEBUG_OPTION --inspect-brk node_modules/jest/bin/jest --runInBand --detectOpenHandles --verbose --config ./jest_ui.config.js",
+		"appium:setup": "output=`appium driver list --installed --json` && (echo $output | grep -q 'uiautomator2' && echo 'uiautomator2 is installed, skipping installation.' || (echo 'uiautomator2 not found, installing...' && appium driver install uiautomator2)) && (echo $output | grep -q 'xcuitest' && echo 'xcuitest is installed, skipping installation.' || (echo 'xcuitest not found, installing...' &&  ...
```
This is nice. I noticed this output when I first ran the commands. 

I wonder if we should look to follow the lead of the `wp-scripts` package an place these larger scripts somewhere. We don't need to necessarily do that now — just "thinking out loud."

geriux replied:
That's a great suggestion! We have some scripts within the `bin` [folder](https://github.com/WordPress/gutenberg/tree/f0b4192b33fc69fe42929ccceda161979435ebc5/packages/react-native-editor/bin) we could move it there and just call the file from here. What do you think?

dcalhoun replied:
Yeah, placing them in the `bin` folder would be a big improvement over capturing this amount of scripting in `package.json`. It will also make reuse far easier, likely leading to fewer bugs.

geriux replied:
I've addressed this in https://github.com/WordPress/gutenberg/pull/55257

dcalhoun reviewed (commented):
Thank you for setting this up! 

I recognize this is a WIP, leaving a few thoughts for consideration.

Commit:
E2E Utils - Revert removing passing the auth token for the jobURL, it also updates the logic to the latest implementation from SauceLabs docs.

Commit:
Remove idleTimeout desktop-only SauceLabs option

Commit:
[Mobile] - Update E2E Block Insertion tests (#55167)

* E2E Utils, updates the typing helpers:

- Removes the platform specfic typing functionality in favor of WebdriverIO's addValue

* E2E Utils - Updates getKeycode and pressKeycode to WebdriverIO

* E2E Utils - Update tapStatusBariOS to WebdriverIO

* E2E Utils - Update toggleHtmlMode to WebdriverIO

* E2E Utils - Update waitForVisible to WebdriverIO

* E2E Editor Page - Update getBlockAtPosition to WebdriverIO

* E2E Editor Page - Update getTitleElement to WebdriverIO

* E2E Editor Page - Update getHtmlContent to WebdriverIO

* E2E Editor Page - Update swipeToolbarToElement to WebdriverIO

* E2E Editor Page - Update addNewBlock to WebdriverIO

* E2E Editor Page - Update findBlockButton to WebdriverIO

* E2E Editor Page - Update assertSlashInserterPresent to WebdriverIO

* E2E EditorPage - Update waitForElementToBeDisplayedById to WebdriverIO

* E2E Test - Block Insertion - Update to WebdriverIO

* Update typeString scroll delay comment

Co-authored-by: David Calhoun <github@davidcalhoun.me>

---------

Co-authored-by: David Calhoun <github@davidcalhoun.me>

dcalhoun committed:
test: Update Search tests for Appium 2 (#55170)

* test: Update getSearchBlockTextElement query

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update waitForElementToBeDisplayedByXPath query

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update verifySearchElementText query

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update getSearchBlockTextElement query

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

Commit:
E2E Editor Page - Update dismissKeyboard to use WebdiverIO's hideKeyboard (#55211)

dcalhoun committed:
test: Update Media Blocks tests for Appium 2 (#55222)

* test: Replace sleep with pause utility

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Replace elementsByAccessibilityId usage

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Replace elementByXPath usage

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Repair enterCaptionToSelectedImageBlock utility

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

Additionally, the `clearValue` method did not succeed on the caption
Button element. Hence, the utility was expanded to instead type into the
TextView.

dcalhoun committed:
test: Update Paragraph tests for Appium 2 (#55215)

* test: Update clickMiddleOfElement helper

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update clickBeginningOfElement helper

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update getTextForParagraphBlockAtPosition helper

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Replace elementsByXPath usage

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

Commit:
E2E Utils - Migrate dragAndDropAfterElement to WebdriverIO (#55254)

Commit:
[Mobile] - Move Appium drivers setup to bash script (#55257)

* Move Appium drivers setup to bash script

* Update drivers' names casing for consistency

* Rename appium:setup to test:e2e:setup

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/appium-2

dcalhoun committed:
test: Update E2E utilities (#55322)

* test: Set static device pixel ratio

WebdriverIO (the replacement for the deprecated WD driver library) does
not have a method for retrieving a device's pixel ratio. Therefore, this
sets an explicit value for each of our test devices.

* test: Update the `getAddBlockButton` utility

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

* test: Update toggleDarkMode utility

Seems this method has changed or been removed for Android.

* test: Increase robustness of waitForKeyboardToBeHidden utility

This utility appeared to fail occasionally due to an empty `addButton`
value.

* test: toggleDarkMode uses consistent approach across platforms

This updates the iOS variant to mirror the structure of the Android
variant in the name of consistency.

* refactor: Rename devicePixelRatio to pixelRatio

Mirror Android implementation found in Appium 2.

* test: Increase logging levels

Attempt to debug CI failures.

```
FAIL __device-tests__/gutenberg-editor-rendering-media-blocks.test.js (73.496 s)
  ● Gutenberg Editor Rendering Media Blocks test › should be able to render blocks correctly

    unknown method: Not implemented

      at getErrorFromResponseBody (file:/Users/runner/work/gutenberg/gutenberg/node_modules/webdriver/build/utils.js:194:12)
      at NodeJSRequest._request (file:/Users/runner/work/gutenberg/gutenberg/node_modules/webdriver/build/request/index.js:164:23)
      at Browser.wrapCommandFn (file:/Users/runner/work/gutenberg/gutenberg/node_modules/@wdio/utils/build/shim.js:81:29)
```

* refactor: Prefer `action` to `touchAction`

Avoid the deprecated `touchAction` command.

> The `touchAction` command is deprecated and will be removed in a future version. We recommend to use the action command instead with pointer type touch, e.g.:
>
> await browser.action('pointer', {
>   parameters: { pointerType: 'touch' }
> })

https://webdriver.io/docs/api/element/touchAction/

This may also relate to the fact that Mobile JSON Wire Protocol is
deprecated as well.

> Mobile JSON Wire Protocol
>
> The Mobile JSON Wire Protocol is a super set of mobile commands on top of the JSON Wire Protocol. Given this one is deprecated the Mobile JSON Wire Protocol also got deprecated. Appium might still support some of its commands but it is not recommended to use them.

https://webdriver.io/docs/api/protocols#mobile-json-wire-protocol

* Revert "test: Increase logging levels"

This reverts commit 7bb46decef07d4ff3c4956d717e291a2d8222b8e.

* test: Set default `swipeFromTo` delay value

This prevents `pause` from being called without a parameter, which
throws an error.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/appium-2

dcalhoun committed:
test: Replace swipe gesture with simple tap (#55427)

It was unclear as to why the swipe gesture failed to close the bottom
sheet, but a simple tap towards the middle-top of the display achieves
the same result.

Commit:
[Mobile] - Appium 2 Migration - Updating more utils (#55420)

* Updates the longPressMiddleOfElement util by using WebdriverIO's action. It adds tapping on the element first before long-pressing due to an issue with iOS not triggering the long-press event. The default waitTime was reduced as well to 1000.

* Update selectTextFromElement to WebdriverIO. It also changes the logic for iOS where it uses the "Select All" option from the context menu instead of using "doubleTap". Timeouts have been reduced as well.

* Refactors selectTextFromElement for Android, the previous implementation would only select the word that was located in the middle of the element, with the new changes it will select all text by long-pressing from the start of the element to the end. Unfortunately we can't access the context menu as we do on iOS so it has to be done manually.

* Add inline duration value and adds inline comments to explain the functionality

Commit:
[Mobile] - Update closePicker util and updates Image Block test data (#55484)

* Update closePicker util name to closeMediaPicker, it also unifies the logic for both platforms by having the same locator and using WebdriverIO's waitForDisplayed.

* Update test data for the image block, by changing the image caption to avoid issues in CI and autocorrect

Commit:
[Mobile] Improve Screen recording on Android for Appium 2 (#55443)

* Improve Screenrecording on Android by making sure of the following:

The sdcard folder should exist in the emulator, if not there's a script that will create it.

For skipped tests, it will remove empty folders that weren't needed.

When having multiple Android emulators opened, it will use the deviceID that Appium is currently using.

* Rename getAndroidDeviceID function to getDeviceID since it's not platform specific

* Fix comment

* Temp: log session

* Use adb devices list to get the first active android emulator

* Moves getAndroidEmulatorID to its own file and replaces its usage. It also adds specifying the device ID for Android local builds for cases when multiple devices/emulators are connected.

* Remove unneeded filtering

* Move setting the androidDeviceID value for Android cases only

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/appium-2

dcalhoun committed:
docs: Update E2E test documentation (#55519)

* docs: Rewrite end-to-end testing documentation

Update for breaking changes from upgrading Appium 2. Reduce verbosity
and duplication to improve the approachability and maintainability of
the documentation.

* docs: Remove E2E test contributing documentation

This was replaced by a section in the directory's README documentation.

* docs: Rewrite end-to-end testing Getting Started section

Update for breaking changes from upgrading Appium 2. Reduce verbosity
and duplication between the `WordPress/gutenberg` repository to improve
the approachability and maintainability of the documentation.

* docs: Update links in Gutenberg for Mobile Apps code documentation

Documents have been updated and consolidated to reduce verbosity
and duplication to improve the approachability and maintainability of
the documentation.

Commit:
[Mobile] Remove `wd` usage and removes disabled tests (#55541)

* Device tests - Removes skipped tests that haven't been enabled back for a while. These were disabled due to flakiness most likely to the interactions with the context menu to copy/paste text. If needed this can be restored once the flakiness issues are worked on.

* Device Tests - Drag & Drop: removes disabled tests. These used utils that were flaky and have been disabled for a long time, they can be added back once stable utils are in place for pasting content.

* Editor Page - Removes setHtmlContent since this util is not being used anymore. It can be added back if its needed.

* Device Tests Utils - Removes usage of the wd library. It also removes, doubleTap, clickElementOutsideOfTextInput, and related utils to the context menu due to them being flaky and not currenlty in use: tapSelectAllAboveElement, tapCopyAboveElement and tapPasteAboveElement.

Commit:
Merge branch 'trunk' into rnmobile/appium-2

dcalhoun committed:
test: Improve E2E test consistency (#55554)

* test: Mitigate conflicts between development server caches

Occasionally, cryptic npm package dependency errors can occur between
test runs. They appear to originate from stale development caches, e.g.
Metro server. Clearing these caches, which E2E tests should not rely
upon mitigates the likelihood of these errors.

* test: Prefix informative E2E test setup logs

Merely an attempt to add structure to the outputted logs.

* test: Avoid setup scripts during test runs

There may be clarity and efficiency by having an explicit setup script.
It was avoid running unnecessary scripts during test runs. It is also
referenced in the current documentation as a script to run
intentionally.

dcalhoun committed:
build: Ignore generated Android files (#55569)

These assets are generated at build time for E2E testing. Seemingly, we
do not leverage these generated files for the binary/bundle that we
provide to the host apps. Therefore, we can safely ignore these files.

geriux commented on packages/sync/src/y-webrtc/crypto.js:
I don't know why these files are being formatted in this PR, I guess we can keep it.

dcalhoun replied:
Yeah, it is a bit bizarre the unformatted files were merged into `trunk`. 🤷🏻 The formatting changes seem appropriate.

geriux reviewed (commented):
I can't approve my own PR but it looks good to me! 🚀

dcalhoun reviewed (approved):
Approved pending a post providing migration details to contributors.

dcalhoun committed:
docs: Avoid GMA project name until it is approved

Use of the "Gutenberg for Mobile Apps" name is still under review. In
the meantime, we should use the existing "Mobile Gutenberg" project
name.
